File: generated_code/cpp/solution_3813_1.cpp
#include <string>
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
private:
    // Helper function to calculate factorials up to n safely.
    // Given string length constraints typically fit within standard limits, 
    // but using long long to prevent overflow during multinomial coefficient calculation.
    vector<long long> getFactorials(int n) {
        vector<long long> fact(n + 1, 1);
        for (int i = 2; i <= n; ++i) {
            // Cap at a large value if it exceeds maximum possible k to prevent overflow
            if (fact[i - 1] > 2e18 / i) {
                fact[i] = 2e18; 
            } else {
                fact[i] = fact[i - 1] * i;
            }
        }
        return fact;
    }

    // Counts the number of distinct permutations possible with the remaining character frequencies
    long long countPermutations(const vector<int>& counts, const vector<long long>& fact) {
        int total = 0;
        for (int c : counts) {
            total += c;
        }
        long long num = fact[total];
        for (int c : counts) {
            if (c > 1) {
                num /= fact[c];
            }
        }
        return num;
    }

public:
    string smallestPalindrome(string s, int k) {
        vector<int> counts(26, 0);
        for (char c : s) {
            counts[c - 'a']++;
        }

        // Validate if a palindrome can be formed
        int odd_count = 0;
        char odd_char = 0;
        for (int i = 0; i < 26; ++i) {
            if (counts[i] % 2 != 0) {
                odd_count++;
                odd_char = 'a' + i;
            }
        }

        // A palindrome can have at most one character with an odd frequency
        if (odd_count > 1) return "";

        // Reduce frequencies to represent only the first half of the palindrome
        vector<int> half_counts(26, 0);
        int half_len = 0;
        for (int i = 0; i < 26; ++i) {
            half_counts[i] = counts[i] / 2;
            half_len += half_counts[i];
        }

        vector<long long> fact = getFactorials(half_len);

        // Check if total possible palindromic permutations is less than k
        long long total_perms = countPermutations(half_counts, fact);
        if (total_perms < k) return "";

        // Construct the first half lexicographically using a next-candidate approach
        string first_half = "";
        long long current_k = k;

        for (int i = 0; i < half_len; ++i) {
            for (int j = 0; j < 26; ++j) {
                if (half_counts[j] > 0) {
                    // Try placing character 'a' + j at the current position
                    half_counts[j]--;
                    long long perms = countPermutations(half_counts, fact);
                    
                    if (current_k <= perms) {
                        // The k-th permutation lies within this character's branch
                        first_half += (char)('a' + j);
                        break;
                    } else {
                        // Skip all permutations starting with this character
                        current_k -= perms;
                        half_counts[j]++; // Backtrack
                    }
                }
            }
        }

        // Form the full palindrome from the first half, the odd middle character (if any),
        // and the reversed first half.
        string second_half = first_half;
        reverse(second_half.begin(), second_half.end());
        
        if (odd_count == 1) {
            return first_half + odd_char + second_half;
        }
        return first_half + second_half;
    }
};

File: generated_code/python/solution_3813_1.py
from typing import List, Dict
from collections import Counter
import math

class Solution:
    def smallestPalindrome(self, s: str, k: int) -> str:
        # Count frequencies of each character
        counts = Counter(s)
        
        # A palindrome can have at most one character with an odd count
        odd_char = ""
        half_counts = {}
        for char, count in counts.items():
            if count % 2 != 0:
                if odd_char:
                    return "" # More than one odd character means no palindrome can be formed
                odd_char = char
            half_counts[char] = count // 2
            
        # Get the unique characters sorted lexicographically for the first half
        unique_chars = sorted(half_counts.keys())
        
        # Calculate the total number of permutations possible for the half-string
        # Formula: total_length! / (count1! * count2! * ...)
        half_length = sum(half_counts.values())
        
        def get_total_permutations(freq: Dict[str, int], total_len: int) -> int:
            ans = math.factorial(total_len)
            for count in freq.values():
                ans //= math.factorial(count)
            return ans
            
        total_perms = get_total_permutations(half_counts, half_length)
        if k > total_perms:
            return "" # Fewer than k distinct palindromic permutations exist
            
        # Reconstruct the first half lexicographically character by character (Digit DP / Combinatorics approach)
        first_half = []
        remaining_len = half_length
        k_remaining = k
        
        for i in range(half_length):
            for char in unique_chars:
                if half_counts[char] > 0:
                    # If we pick 'char' for the current position, calculate how many permutations 
                    # can be formed with the remaining characters
                    half_counts[char] -= 1
                    perms_with_char = get_total_permutations(half_counts, remaining_len - 1)
                    
                    if k_remaining <= perms_with_char:
                        # The k-th permutation lies within the combinations starting with 'char'
                        first_half.append(char)
                        remaining_len -= 1
                        break
                    else:
                        # Skip all permutations starting with 'char' and look into the next character
                        k_remaining -= perms_with_char
                        half_counts[char] += 1 # Backtrack
                        
        # Construct the final full palindrome
        first_half_str = "".join(first_half)
        return first_half_str + odd_char + first_half_str[::-1]